From 3d20682cb9c70ab72c8ec832026d1b65dd86aeea Mon Sep 17 00:00:00 2001
From: Douglas <douglas.macfarlane@wolfsonmicro.com>
Date: Fri, 22 Nov 2013 15:04:38 +0000
Subject: [PATCH 11/14] ASoC: WM8804: MCLK_DIV to be set by set_clkdiv rather
 than within set_pll

This change allows one to deal with any conflicts which
can arise as MCLK_DIV is in the path for the TX clock,
but not the RX (it takes the pll output directly).

Change-Id: If0170d35568b75c47a30163bfb747ad202f7cad4
Signed-off-by: Douglas MacFarlane <douglas.macfarlane@wolfsonmicro.com>
---
 sound/soc/codecs/wm8804.c |   22 ++++++++++------------
 sound/soc/codecs/wm8804.h |    1 +
 2 files changed, 11 insertions(+), 12 deletions(-)

diff --git a/sound/soc/codecs/wm8804.c b/sound/soc/codecs/wm8804.c
index 08aab21..704f98c 100644
--- a/sound/soc/codecs/wm8804.c
+++ b/sound/soc/codecs/wm8804.c
@@ -310,16 +310,11 @@ struct pll_div {
 static struct {
 	unsigned int div;
 	unsigned int freqmode;
-	unsigned int mclkdiv;
 } post_table[] = {
-	{  2,  0, 0 },
-	{  4,  0, 1 },
-	{  4,  1, 0 },
-	{  8,  1, 1 },
-	{  8,  2, 0 },
-	{ 16,  2, 1 },
-	{ 12,  3, 0 },
-	{ 24,  3, 1 }
+	{  2,  0 },
+	{  4,  1 },
+	{  8,  2 },
+	{ 12,  3 }
 };
 
 #define FIXED_PLL_SIZE ((1ULL << 22) * 10)
@@ -338,7 +333,6 @@ static int pll_factors(struct pll_div *pll_div, unsigned int target,
 		tmp = target * post_table[i].div;
 		if (tmp >= 90000000 && tmp <= 100000000) {
 			pll_div->freqmode = post_table[i].freqmode;
-			pll_div->mclkdiv = post_table[i].mclkdiv;
 			target *= post_table[i].div;
 			break;
 		}
@@ -405,8 +399,7 @@ static int wm8804_set_pll(struct snd_soc_dai *dai, int pll_id,
 		snd_soc_update_bits(codec, WM8804_PLL4, 0xf | 0x10,
 				    pll_div.n | (pll_div.prescale << 4));
 		/* set mclkdiv and freqmode */
-		snd_soc_update_bits(codec, WM8804_PLL5, 0x3 | 0x8,
-				    pll_div.freqmode | (pll_div.mclkdiv << 3));
+		snd_soc_update_bits(codec, WM8804_PLL5, 0x3, pll_div.freqmode);
 		/* set PLLK */
 		snd_soc_write(codec, WM8804_PLL1, pll_div.k & 0xff);
 		snd_soc_write(codec, WM8804_PLL2, (pll_div.k >> 8) & 0xff);
@@ -465,6 +458,11 @@ static int wm8804_set_clkdiv(struct snd_soc_dai *dai,
 		snd_soc_update_bits(codec, WM8804_PLL5, 0x30,
 				    (div & 0x3) << 4);
 		break;
+	case WM8804_MCLK_DIV:
+		snd_soc_update_bits(codec, WM8804_PLL5, 0x8,
+				    (div & 0x1) << 3);
+		break;
+
 	default:
 		dev_err(dai->dev, "Unknown clock divider: %d\n", div_id);
 		return -EINVAL;
diff --git a/sound/soc/codecs/wm8804.h b/sound/soc/codecs/wm8804.h
index 8ec14f5..a55612a 100644
--- a/sound/soc/codecs/wm8804.h
+++ b/sound/soc/codecs/wm8804.h
@@ -57,5 +57,6 @@
 #define WM8804_CLKOUT_SRC_OSCCLK		4
 
 #define WM8804_CLKOUT_DIV			1
+#define WM8804_MCLK_DIV	 		        2
 
 #endif  /* _WM8804_H */
-- 
1.7.9.5

